### PR TITLE
Adding sampler write and read functions to sampler.py 

### PR DESCRIPTION
**Pull request by han94ros**
# ARTIQ Pull Request

## Description of Changes
Adds sampler_mu_write and sampler_mu_read functions to the sampler core device driver. The purpose is to allow for control of the buffering of samples e.g sampling on a timestamp and then recovering the results at a later time. 
Addresses issue #1738.

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |

### All Pull Requests

- [x] Use correct spelling and grammar.
- [x] Update [RELEASE_NOTES.rst](../RELEASE_NOTES.rst) if there are noteworthy changes, especially if there are changes to existing APIs.
- [x] Close/update issues.
- [ ] Check the copyright situation of your changes and sign off your patches (`git commit --signoff`, see [copyright](../CONTRIBUTING.rst#copyright-and-sign-off)).

### Code Changes

- [ ] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [x] Test your changes or have someone test them. Mention what was tested and how.
- [x] Add and check docstrings and comments
- [ ] Check, test, and update the [unittests in /artiq/test/](../artiq/test/) or [gateware simulations in /artiq/gateware/test](../artiq/gateware/test)

### Documentation Changes

- [x] Check, test, and update the documentation in [doc/](../doc/). Build documentation (`cd doc/manual/; make html`) to ensure no errors.

### Git Logistics

- [ ] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

File: artiq/coredevice/sampler.py [hunk 1]
```diff
@@ -129,6 +129,32 @@ def sample_mu(self, data):
             data[i] = val >> 16
             val &= 0xffff
             data[i - 1] = -(val & mask) + (val & ~mask)
+            
+
```
Comment (jordens) [inline artiq/coredevice/sampler.py:133]: There are a bunch of checkboxes that you haven't completed yet. One of them is style. Please make some effort to propose quality code that matches the prevalent style and orthography.

File: artiq/coredevice/sampler.py [hunk 2]
```diff
@@ -129,6 +129,32 @@ def sample_mu(self, data):
             data[i] = val >> 16
             val &= 0xffff
             data[i - 1] = -(val & mask) + (val & ~mask)
+            
+    
+    @kernel
+    def sample_mu_write(self, data):
```
Comment (jordens) [inline artiq/coredevice/sampler.py:135]: Then only pass a numer, not a list.

Comment (jordens) [inline artiq/coredevice/sampler.py:135]: This method should then be used in `sample_mu()`.

File: artiq/coredevice/sampler.py [hunk 3]
```diff
@@ -129,6 +129,32 @@ def sample_mu(self, data):
...
+        """prepare ADC.
```
Comment (jordens) [inline artiq/coredevice/sampler.py:136]: Please write accurate and quality comments. They will appear in the documentation. "prepare ADC" isn't what this does at all.

File: artiq/coredevice/sampler.py [hunk 4]
```diff
@@ -129,6 +129,32 @@ def sample_mu(self, data):
...
+
+        Writing appropriate data to shift register and start transfer
```
Comment (jordens) [inline artiq/coredevice/sampler.py:138]: Please write entire sentences.

File: artiq/coredevice/sampler.py [hunk 5]
```diff
@@ -129,6 +129,32 @@ def sample_mu(self, data):
...
+        self.cnv.pulse(30*ns)  # t_CNVH
+        delay(450*ns)  # t_CONV
+        mask = 1 << 15
+        for i in range(len(data)//2):
+            self.bus_adc.write(0)
+
+            
+        
+    @kernel
+    def sample_mu_read(self, data):
```
Comment (jordens) [inline artiq/coredevice/sampler.py:149]: This method should then also be used in `sample_mu()`.

File: artiq/coredevice/sampler.py [hunk 6]
```diff
@@ -129,6 +129,32 @@ def sample_mu(self, data):
...
+        """read in the values submitted to the SPI core.
```
Comment (jordens) [inline artiq/coredevice/sampler.py:150]: Please write proper docstrings with proper sentences.

**Review by jordens — CHANGES_REQUESTED** [review]
This approach is OK. Just need to meet the minimum requirements.
And please show that you have tested this code.

**Comment by han94ros** [comment]
Can you please reopen this PR so that I can finish the tasks.

**Pull request closed without merging**